service/src/main/java/bio/terra/workspace/db/FolderDao.java

```diff
@@ -0,0 +1,254 @@
+        throw new DuplicateFolderDisplayNameException(
+            String.format("Folder with display name %s already exists", folder.displayName()));
+      }
+      throw e;
+    } catch (DataIntegrityViolationException e) {
+      if (e.getMessage() != null
+          && e.getMessage().contains("violates foreign key constraint \"fk_folder_wid\"")) {
+        throw new WorkspaceNotFoundException(
+            String.format(
+                "Failed to find workspace %s in which to create the folder", folder.workspaceId()));
+      }
+      throw e;
+    }
+  }
+
+  @WriteTransaction
+  public boolean updateFolder(
+      UUID workspaceId,
+      UUID folderId,
+      @Nullable String displayName,
+      @Nullable String description,
+      @Nullable UUID parentFolderId,
+      @Nullable Boolean updateParent) {
+    if (displayName == null
+        && description == null
+        && parentFolderId == null
+        && updateParent == null) {
+      throw new MissingRequiredFieldException("Must specified fields to update");
+    }
+    if (parentFolderId != null) {
+      if (getFolderIfExists(workspaceId, folderId).isEmpty()) {
+        throw new FolderNotFoundException(
+            String.format(
+                "Cannot update parent folder to %s because it is not found in workspace %s",
+                parentFolderId, workspaceId));
+      }
+    }
+    if (canFormCyclicCycle(workspaceId, parentFolderId, folderId)) {
+      throw new BadRequestException(
+          String.format(
+              "Cannot update parent folder id to %s as it will create a cycle", parentFolderId));
+    }
+    var params = new MapSqlParameterSource();
+    Optional.ofNullable(displayName).ifPresent(name -> params.addValue("display_name", name));
+    Optional.ofNullable(description).ifPresent(d -> params.addValue("description", d));
+    if (Boolean.TRUE.equals(updateParent)) {
+      params.addValue(
+          "parent_folder_id",
+          Optional.ofNullable(parentFolderId).map(UUID::toString).orElse(DEFAULT_ROOT));
+    }
+    StringBuilder sb = new StringBuilder("UPDATE folder SET ");
+
+    sb.append(DbUtils.setColumnsClause(params));
+
+    sb.append(" WHERE workspace_id = :workspace_id AND id = :folder_id");
+    params
+        .addValue("workspace_id", workspaceId.toString())
+        .addValue("folder_id", folderId.toString());
+    try {
+      int rowsAffected = jdbcTemplate.update(sb.toString(), params);
+      return rowsAffected > 0;
+    } catch (DuplicateKeyException e) {
+      if (e.getMessage() != null
+          && e.getMessage()
+              .contains(
+                  "duplicate key value violates unique constraint \"folder_display_name_parent_folder_id_workspace_id_key\"")) {
+        throw new DuplicateFolderDisplayNameException(
+            String.format(
+                "Fails to update due to duplicate display name at the same folder level"));
```

Comment:
Folder with display name %s already exists in parent folder %s

Reply:
it's because display name can be null and I dont' want to query the database for the folder display name just to how error message.

For example, there is a foo and another foo under foo.
If I update foo to be the same level as foo, the update request will have display name null but the request will fail because there is already a foo at the top level.